src/self-update/README.md

```diff
@@ -0,0 +1,89 @@
+# Introduction
+
+That document specifies the feature self-update for thin-edge.
+
+The self-update allows a Cloud Operator to update one thin-edge version that is installed on a device to a newer version, via the Cloud connection. The feature is referred as self-update cause that update process is managed by thin-edge own's Software Management.
+
+# Requirements
+
+* Be default self-update shall use Debian Package Manager APT.
+  * NOTE: APT Package Manager provides a very basic rollback mechanism that still can fail (e.g. due to some unexpected device state or power-loss).
+* Device Owner shall be able to use a power-fail safe filesystem and transaction-safe package manager (instead of APT) - if the device support any.
+  * Therefore an extensible plugin shall be provided.
+* In thin-edge package's installation logic no external tools (e.g. grep, sed, ...) shall be used. Instead Rust code/libraries shall be used.
+  * This is to reduce external dependencies and so reduce risc for failures due missing tools or wrong version of tools.
+
+# Expected Pre-Conditions 
+
+The device and it's thereon installed thin-edge has to follow precondition below to allow successful self-update:
+
+1) The current (old) thin-edge on the device must be installed with usual thin-edge installation, or at leat as usual thin-edge installation would do. That is to assure new thin-edge version that is about to be installed has all external and internal files, folders and resources (e.g. config files) available.
+2) Deployment of thin-edge and mosquitto must not be changed manually after initial installation (e.g. moving/renaming installed files, moving/renaming configuration files used by thin-edge, renaming users/groups created during installation, ...).
+
+3) Package Dependencies need to be solved using SW Management before self-update. This is since self-update will disable auto-dep solving to reduce additional risk during self-update.
+
+*Hint: Before executing self-update on devices in the field it shall be tested on a kind of golden device that equals the device(s) in the field in system-environment, configuration and thin-edge version.*
+
+
+# Execution Flow and involved Components
+![Sequence Diagram Update SW-list](images/self-update.drawio.svg)
+
+
+# Design Principles
+
+### Details about the SM Plugin
+
+* A designated Software Management module type - namely "tedge" - will be used manage the self-update.
+
+* The existing SM plugin "apt" will be used and extended to manage that module type "tedge". 
+  * A device owner can implement an own plugin for module type "tedge", e.g. to use any other package manager than APT for self-update. 
+
+* A softlink to be created in `/etc/tedge/sm-plugins/` named `tedge` that points to `apt` plugin, will make the SM agent aware.
```

Comment:
Since the `tedge` plugin is just a symlink of `apt` plugin, how will we prevent the tedge plugin from re-reporting the apt software list(coming from the symlinked apt plugin) when the agent queries it for the software list? We'll make the tedge-agent exclude the tedge plugin from reporting the software list ? In that case, how will the tedge binaries have the `::tedge` suffix when reported in the UI? Adding that suffix from inside the apt plugin(e.g `tedge-agent::tedge`) wouldn't work as it will be further suffixed with `::apt` by the agent before sending to the cloud.

Reply:
Got the answer to this from the #759 which proposes the apt plugin to detect whether the plugin is called as `tedge` or as `apt` and then report the list accordingly.